db/user.go

```diff
@@ -221,3 +222,71 @@ func SyncUsers(course *model.Course, newUsers map[string]*model.User,
 
     return syncResult, nil;
 }
+
+// ResolveUsers is a helper function to maps string representations of roles and * (all roles) to the emails for users with those roles.
+// The function takes an optional course and a list of strings, containing emails, roles, and * as input and returns an alphanumerically sorted slice of lowercase emails without duplicates.
+// A fatal error occurs if a course cannot be opened and the input contains roles or a *.
+// A warning will be logged for every invalid role.
+func ResolveUsers(course *model.Course, emails []string) ([]string, error) {
+    if (backend == nil) {
+        return nil, fmt.Errorf("Database has not been opened.");
+    }
+
+    emailSet := map[string]any{};
+    roleSet := map[string]any{};
+
+    // Iterate over every given string, checking for emails, roles, and *, which denotes all users.
+    for _, email := range emails {
+        email = strings.ToLower(strings.TrimSpace(email));
+        if (email == "") {
+            continue;
+        }
+
+        if (strings.Contains(email, "@")) {
+            emailSet[email] = nil;
+        } else {
+            if (email == "*") {
+                allRoles := model.GetAllRoleStrings();
+                for role := range allRoles {
+                    roleSet[role] = nil;
+                }
+            } else {
+                roleSet[email] = nil;
+            }
+        }
+    }
+
+    if (len(roleSet) > 0) {
+        users, err := GetUsers(course);
+        if (err != nil) {
+            return nil, err;
+        }
+
+        for roleString := range roleSet {
+            if (roleString == "unknown") {
+                continue;
+            }
+
+            if (model.GetRole(roleString) == model.RoleUnknown) {
+                log.Warn("Invalid role given to ResolveUsers.", course, log.NewAttr("role", roleString))
+            }
+        }
+
+        for _, user := range users {
+            // Add a user if their role is set.
+            _, ok := roleSet[model.GetRoleString(user.Role)];
```

Comment:
I'm not a huge fan of checking against the string roles map that still may have junk in it.
I would say that you should either build up a map/set of model.UserRole in the previous loop that you can check against,
or delete the bad entries where you warn in the previous loop.